core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala

```diff
@@ -477,14 +475,25 @@ class CollectorService(
     * Creates an Access-Control-Allow-Origin header which specifically allows the domain which made
     * the request
     * @param request Incoming request
-    * @return Header allowing only the domain which made the request or everything
+    * @param domainRestriction A list of domains which the ACAO header will only be served for, or `None`
+    * if there are no restrictions.
+    * @return A singleton list containing the specified header. If it violates the restrictions, the list
+    * will be empty. If the request had no `Origin` header, the response header will permit every domain.
     */
-  def accessControlAllowOriginHeader(request: HttpRequest): HttpHeader =
-    `Access-Control-Allow-Origin`(request.headers.find {
-      case `Origin`(_) => true
-      case _           => false
+  def accessControlAllowOriginHeader(
+    request: HttpRequest,
+    restrictDomainsTo: Option[List[String]],
+  ): List[HttpHeader] =
+    request.headers.collectFirst {
+      case `Origin`(origins) => origins
     } match {
-      case Some(`Origin`(origin)) => HttpOriginRange.Default(origin)
-      case _                      => HttpOriginRange.`*`
-    })
+      case None         => List(`Access-Control-Allow-Origin`(HttpOriginRange.`*`))
+      case Some(origins) => {
+        val hosts: Seq[String] = origins.map(_.host.value)
+        if (restrictDomainsTo.forall(hosts.intersect(_).nonEmpty))
```

Comment:
Along similar lines like my comment on `case None` above: what is the benefit of this check? The config should be our source of truth as to what domains should be allowed, regardless of whether they are listed in the `Origin` header or not.

Am I missing a use-case here, where we want to further narrow down the list of allowed domains based on origin?

Reply:
According to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin#examples), the HTTP spec doesn't provide a way to respond with multiple domains in the `Access-Control-Allow-Origin` header. Instead, the recommended way is to check against the `Origin` header and respond with only that origin if it is found to be valid. I am not entirely sure why Akka's `HttpOriginRange` class expects a `Seq<>` rather than a single element (probably because that class is expected elsewhere in contexts where that would make sense) but it wouldn't be correct here.

The original code operated under the assumption that the collector should always allow the origin. Thus it returned `*` for the ACAO in cases where there was no origin specified in the request, and if an origin was specified, it would return that for the ACAO (since returning `*` in that case would be illegal).

The goal of the change was for this behavior to be preserved as long as the user does not specify any domain restrictions. If restrictions are specified, things look a little different:

- If there is an `Origin` header, we can check its value against the allowed ones. If it doesn't match, not sending the header is the correct behavior and was [recommended by the original creator of the issue](https://github.com/snowplow/stream-collector/issues/15#issue-631619350). Not sending the header will cause the browser to deny access to the response as a default.

- If there is no `Origin` header, we ignore the domain restrictions and keep returning `*`, just as the original version did. There is nothing we can really do about non-cross-origin requests. Perhaps we could check for the special case where the config specifies exactly one domain and return it in this case, but I'm skeptical of whether that would be worth it.

Hopefully this addresses your comments on the implementation/tests and I haven't missed anything.

Reply:
**Regarding point 1**

> The Origin header can contain multiple domains

Well.... the akka type signature makes it look like that's true.  But I cannot find any evidence or documentation to support that.  My understanding is that browser CORS requests contain a single origin.

> with the changes proposed at the moment, we're still sending back all of the Origin domains

Technically Dilyan is right.... but I see this is a non-issue in reality.  There will only ever be one origin in a browser's CORS request.  (The [example with cookie domains](https://github.com/snowplow/stream-collector/blob/master/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala#L408-L428) is a slightly different case because it is has to match subdomains, not just exact matches).

**Regarding point 2**

> There's an outstanding question what to do with:
> `allowedDomains = [ "first.example.com", "second.example.com" ]`

^^ That is an extremely common use case.  A company often has multiple websites on multiple domains that all want to send events to the same collector.  The desired behaviour should be: Allow origins from _either_ `first.example.com` _or_ `second.example.com`.  Therefore I disagree with the current implementation, which requires all configured domains to match the request domain:

```scala
restrictDomainsTo.forall(hosts.intersect(_).nonEmpty)
```

should be

```scala
restrictDomainsTo.exists(hosts.contains(_))
```

I also disagree with Dilyan's suggestion that we return the first domain from the list of allowed domains.  The correct behaviour is to return the domain from the request header, because that is what the browser expects to receive.